### PR TITLE
config: expand tilde for ~/.ceph/$cluster.conf

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Fixes: http://tracker.ceph.com/issues/23215
Signed-off-by: Rishabh Dave <ridave@redhat.com>

### CODE REVIEW

**Comment by wjwithagen** [comment]
I would very much not do this in application code.
`~` expansion is most certainly a shell type thing.
So best just leave it there.

The fact that it is somewhere in the CEPH config is IMHO a problem in itself.
If anything needs to be done like that I suggest making it expand $home, like things as $cluster, that keeps things consistant

And even then the way it is now implemented is that EVERY `~` is replaced, even if it is not at the beginning of the file-path. Which is way much more than what is done by any shell expansion.

**Comment by rishabh-d-dave** [comment]
> The fact that it is somewhere in the CEPH config is IMHO a problem in itself.
If anything needs to be done like that I suggest making it expand $home, like things as $cluster, that keeps things consistant

Why do you think it is about ceph config?

It is not about tilde somewhere in ceph config. It is with the fact that `~/.ceph/$cluster.conf` is checked and  processed for '$' properly but never for '~'. As a result, files like `~/.ceph/ceph.conf` are never opened by the rados code even if they are meant to be opened [1]. This [2] where the bug is reproduced.

> And even then the way it is now implemented is that EVERY ~ is replaced, even if it is not at the beginning of the file-path. Which is way much more than what is done by any shell expansion.

No. My code specifically looks if tilde is at the zeroth index of the given string `s`. So, not every tilde would be replaced.

[1] https://github.com/ceph/ceph/blob/master/src/common/config.cc#L45
[2] https://github.com/ceph/ceph/pull/20598#issuecomment-370227953

**Comment by wjwithagen** [comment]
Well my argument is that is should not have been used in this format to begin with, and the PR should be to remove `~/.ceph/$cluster.conf` from
```
./src/common/config.cc:44:static const char *CEPH_CONF_FILE_DEFAULT = "$data_dir/config, /etc/ceph/$cluster.conf, ~/.ceph/$cluster.conf, $cluster.conf"
```
And if anything like this was required, replace it by $home/.ceph/$cluster.conf.

File: src/common/config.cc [hunk 1]
```diff
@@ -248,6 +248,9 @@ int md_config_t::parse_config_files(const char *conf_files,
       } else {
 	cfl.erase(p++);  // ignore this item
       }
+    } else if (s.find("~") == 0) {
+	s.replace(s.find("~"), 1, string(getenv("HOME")));
```
Comment (wjwithagen) [inline src/common/config.cc]: Could replace `s.replace(s.find("~"), 1, string(getenv("HOME")));`
by `s.replace(0, 1, string(getenv("HOME")));`

File: src/common/config.cc [hunk 2]
```diff
@@ -248,6 +248,9 @@ int md_config_t::parse_config_files(const char *conf_files,
       } else {
 	cfl.erase(p++);  // ignore this item
       }
+    } else if (s.find("$home") != string::npos) {
+	s.replace(0, 5, string(getenv("HOME")));
```
Comment (rishabh-d-dave) [inline src/common/config.cc]: Would it be more appropriate to move this part to somewhere in expand_meta()?

Comment (wjwithagen) [reply src/common/config.cc]: @rishabh-d-dave 
I think that would make sense, it would then be more universially applicable.

**Comment by liewegas** [comment]
retest this please

**Comment by rishabh-d-dave** [comment]
@wjwithagen I don't why make check failed. Locally, it's passing for me.

**Comment by wjwithagen** [comment]
@rishabh-d-dave 
Looks like ceph-conf fails.
Could be transient, but might have to do with `~` change, which it this case the `$home` translation does not work out the same as in your work environment. 
And that could be that because `~`  was non functional before your change.
But now it is, and the `$home/ceph.conf` does not actually contain what you need in the jenkins test env.

Fixing bug like this, where Jenkins does something different are not easy.
Might want to set to DNM, and submit some debug code, to later remove it once you fixed it.

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-03-14_04:42:43-rados-wip-kefu-testing-2018-03-14-1106-distro-basic-mira/

- SLOW_OPS failures are introduced by #20660
- the coredump is addressed by #20909

**Comment by rishabh-d-dave** [comment]
Before ceph-conf runs this time, the chosen ceph conf file would be displayed and it would be checked if it exists. WRT to how Jenkins runs the test, I'm not sure if the debugging code would really help. 

@wjwithagen 
> Might want to set to DNM, and submit some debug code, to later remove it once you fixed it.

What's DNM?

**Comment by tchaikov** [comment]
@rishabh-d-dave "DNM" stands for "do not merge". @wjwithagen suggested to add this label to your PR so we don't merge it by accident.

**Comment by rishabh-d-dave** [comment]
From the output of my debug code, there is not a single instance where ~/.ceph/ceph.conf was chosen from CEPH_CONF_FILE_DEFAULT. This means that code under this PR is not being tested, so the errors can not be relevant.

File: src/common/config.cc [hunk 3]
```diff
@@ -1156,6 +1169,8 @@ Option::value_t md_config_t::_expand_meta(
 	out += stringify(getpid());
       } else if (var == "cctid") {
 	out += stringify((unsigned long long)this);
+      } else if (var == "home") {
+	out = string(getenv("HOME"));
```
Comment (liewegas) [inline src/common/config.cc]: Small issue: getenv() can return NULL, and std:;string crashes when passed null.  Change to
```
const char *home = getenv("HOME");
out = home ? std::string(home) : std::string();
```
or similar

Comment (rishabh-d-dave) [reply src/common/config.cc]: Done.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-03-27_08:29:32-rados-wip-kefu-testing-2018-03-27-1407-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-03-27_12:22:51-rados-wip-kefu-testing-2018-03-27-1754-distro-basic-smithi/

**Pull request merged**